### PR TITLE
Refactor parsing

### PR DESCRIPTION
**Pull request by timothyaveni**
Lots going on here! The goal of this PR is to do a lot of the initial work required to get Jupytutor ready for Data 8, and also for future DS courses (like Data 140, which we want to get Jupytutor ready for within the next few weeks).

## Parsing refactor

The current code parses the notebook by looking at the HTML rendered into the user's browser, rather than the notebook source code itself (as provided by the JupyterLab extension API). There are pros and cons to this; a big pro is that we don't need to re-parse the Markdown if we need that info, and also that we have a clear view into exactly what outputs are visible to the user. But there were a few issues with this approach:

### Notebook-load race condition

In both dev and prod, I sometimes had trouble getting the plugin to load. I tracked this down to a race condition on notebook parsing:

- there's a 500ms wait once the notebook begins loading, after which we check through the notebook's rendered cell content to find the activation flag (a tag which must be present in some cell). this is also where we look for textbook links.
- I found that it was possible for the activation flag not to be ready yet, and also for the activation flag to be ready but for the textbook links not to be parsed successfully -- in both cases, this was because the render wasn't quick enough for us to grab post-render data.
- I also saw some cases in testing (that I couldn't track down consistently) where certain cells were empty in the prompt to the LLM; I think this was related to the race condition.

It's generally more robust (both for this race-condition reason, but also because the render output is an unstable API compared to the plugin API -- the rendered HTML may change in structure in future versions of Jupyter) to parse from the notebook source.

### Context bug

I think there was a bug with Markdown cells where the LLM response from the last chat would actually get included as part of the "student's FRQ answer" in subsequent prompts (I think because we were getting cell text with `.innerText`, which was capturing the Jupytutor module). It'd be possible to address this specific bug (e.g., with something similar to `removeOutputTextFromInputText` that was in `parseNB.ts` before), but this more general solution should prevent that genre of bug from appearing in the future by keeping the input (notebook data) and output (rendered notebook editor) separate from one another, interacting directly with the notebook's model data from the extension API.

## Move config into notebook-wide metadata

Notebooks have a JSON object storing metadata for the entire notebook, which is great for global settings (like the activation flag) -- this helps us avoid parsing every cell in the notebook to grab global data. I've moved the activation flag into the config, and moved configuration into notebooks.

The main motivation for moving where configuration happens is to make sure that this generalizes into future deployments. We won't be able to deploy Jupytutor into the main `datahub.` instance (which is multi-tenant, used for many DS courses) until it's off-by-default and possible for individual courses to configure. I think the best way to do this is to have the notebooks themselves include the activation information. (students can, of course, flip this activation flag on their own to turn Jupytutor on manually even when the course staff disabled it, but this is essentially impossible to prevent, since users have extensive access to the virtual machine/container running their JupyterHub instance -- and our goal is really only to make it harder than it'd be to ask for the answers from an external LLM tool, anyway.)

I'm not 100% sure the config-load-from-server feature was working in production before; I think the server was giving a 404 for the `config.json`. It's also worth mentioning that the old config format included regexes, which aren't top-level JSON primitives, so they'd probably need to be replaced with string regexes. I plan to clean up this code in a future PR.

We're now using `zod` to parse the config, which gives us TypeScript types to play with, alongside runtime validation that the config loaded from the notebook metadata actually matches the intended type. This also lets us co-locate the config's default values with its type definition (see `schemas/config.ts`).

## Move some logic into the config

To give course staff more fine-grained control, in particular so that Jupytutor can show up only on certain cells, I've added a rule-based system into the config that decides whether Jupytutor appears after a cell is finished executing. It'll need to be documented -- and may need some refinement -- but here's an example configuration (which I think we'll use for Data 8):

```
"rules": [
    {
      "_comment": "Default rule: disable chat in all cells",
      "config": {
        "chatEnabled": false
      }
    },
    {
      "when": {
        "AND": [
          {
            "cellType": "code"
          },
          {
            "hasError": true
          }
        ]
      },
      "config": {
        "chatEnabled": true,
        "chatProactive": true,
        "quickResponses": ["Explain this error."]
      }
    },
    {
      "_comment": "Code cell (likely a grader cell) following an answer cell with all tests passing",
      "when": {
        "AND": [
          {
            "cellType": "code"
          },
          {
            "OR": [
              {
                "nearbyCell": {
                  "relativePosition": -1,
                  "matches": {
                    "tags": {
                      "any": "otter_answer_cell"
                    }
                  }
                }
              },
              {
                "tags": {
                  "any": "jupytutor_grader_cell"
                }
              }
            ]
          },
          {
            "output": {
              "matchesRegex": {
                "pattern": ".*passed!.*",
                "flags": "i"
              }
            }
          }
        ]
      },
      "config": {
        "chatEnabled": true,
        "chatProactive": true,
        "quickResponses": [
          "I still don't feel confident in my answer.",
          "Provide me three important review materials.",
          "Can I make further improvements?"
        ]
      }
    },
    {
      "_comment": "Code cell (likely a grader cell) following an answer cell without all tests passing",
      "when": {
        "AND": [
          {
            "cellType": "code"
          },
          {
            "OR": [
              {
                "nearbyCell": {
                  "relativePosition": -1,
                  "matches": {
                    "tags": {
                      "any": "otter_answer_cell"
                    }
                  }
                }
              },
              {
                "tags": {
                  "any": "jupytutor_grader_cell"
                }
              }
            ]
          },
          {
            "output": {
              "matchesRegex": {
                "pattern": ".*Test case failed.*",
                "flags": "i"
              }
            }
          }
        ]
      },
      "config": {
        "chatEnabled": true,
        "chatProactive": true,
        "quickResponses": [
          "Explain this error.",
          "Provide a concise list of important review materials.",
          "What progress have I made so far?"
        ]
      }
    },
    {
      "_comment": "Markdown answer cell",
      "when": {
        "AND": [
          {
            "cellType": "markdown"
          },
          {
            "tags": {
              "any": "otter_answer_cell"
            }
          }
        ]
      },
      "config": {
        "chatEnabled": true,
        "chatProactive": true,
        "quickResponses": ["Evaluate my answer."]
      }
    }
  ]
```

This config says:
- by default, don't show the plugin
- whenever a code cell has an error, do show it
- whenever a code cell is preceded by `otter_answer_cell` and the output says `passed`, show (with certain pre-configured 'quick responses') -- this is to capture `grader.check` lines
- whenever a code cell is preceded by `otter_answer_cell` and the output says `Test case failed`, show (with different quick responses)
- after editing a markdown cell tagged `otter_answer_cell`, do show it

This is different from the logic before -- and possibly requires a little more babysitting from the course staff editing the notebook, since it doesn't do AST analysis on the grader cells to find them -- but I think the more general configurability is an overall win here. The config allows for some semi-complex logic, including regex matching, matching relative to nearby cells, and tag-based matching (which I think will be a big win for customizability, since course staff will be able to tag anything to override Jupytutor on/off).

This also goes beyond just an enable/disable flag; we can also customize which 'quick responses' appear, and down the road we can also stack 'instructor notes' on cells (either overriding or extending prior prompts). I think this will be a great tool for customization: for example, inserting a note into the LLM prompt based on certain types of autograder failures (like, "don't fixate on `hashlib`, that's just how we check passed/failed tests"), triggered by regexes on the output.

One big change here is that the 'cell type' that comes from parsing is _much_ more basic -- it's just `code`, `markdown`, or `unknown`. Instead of hard-coding pre-defined behavior based on a number of cell types (like 'grader passed/failed/not found'), the parse itself only fetches enough information to match to these config rules (so, instead of 'success' as a cell type, 'success_regex' to detect success, and 'show_on_success' as a boolean config flag, we just have a mapping of 'regex match rule' -> 'config override', which allows config authors to implement the same behavior).

## Misc changes

I'm curious (cc @kevyg03) - what was the 'active index correction' doing? It seems like this was related to the auto-advance that happens when a cell is executed, possibly related to output areas? I didn't reintroduce this in the new version of the parsing code, because it didn't seem necessary, but I want to make sure I haven't missed anything. I think this is used by local context correction? I'm hoping to revisit that code next, but even if this is off-by-one, hopefully it doesn't cause big issues in the short-term. Just don't want to lose track of this.

This removes the custom widget for the `grader_not_initialized` case, since we don't look explicitly for `grader.check` anymore. This was a cool widget, but I think it's best to avoid letting the scope of the plugin stretch too much, so that it doesn't end up as too much of a 'kitchen sink' feature-set, which would make it harder for future users to develop a consistent mental model of what the plugin provides. Providing course-specific guidance (external to the AI help) is a good idea, but it's best left to a separate plugin, or even just student-facing documentation written in the ntoebook.

I've added some `.nix` files so that I can get a (somewhat) reproducible development environment working on my machines. Safe to ignore these, if the development environment still works for you!

### CODE REVIEW

File: src/Jupytutor.tsx [hunk 1]
```diff
@@ -114,10 +144,7 @@ export const Jupytutor = (props: JupytutorProps): JSX.Element => {
           role: 'system' as const,
           content: [
             {
-              text:
```
Comment (timothyaveni) [inline src/Jupytutor.tsx]: Planning to revamp some of this context construction in a future PR; a lot will probably come from config

File: src/Jupytutor.tsx [hunk 2]
```diff
@@ -616,42 +653,6 @@ export const Jupytutor = (props: JupytutorProps): JSX.Element => {
...
-        {
-          id: 'improvements',
-          text: 'Can I make further improvements?'
-        }
-      ];
-    else opts = [];
-
-    return opts;
-  }, [cellType]);
-
```
Comment (timothyaveni) [inline src/Jupytutor.tsx]: this is in config now

File: src/config.ts [hunk 3]
```diff
@@ -1,23 +1,21 @@
...
+
 /**
  * THIS IS THE DEFAULT CONFIGURATION FOR THE EXTENSION.
  *
  * It is overridden by the config file in ~/.config/jupytutor/config.json
  *
  * Structure ~/.config/jupytutor/config.json the same as the exported config object.
  */
-export const config = {
+export const defaultConfig: z.output<typeof ConfigSchema> = {
```
Comment (timothyaveni) [inline src/config.ts]: this is not actually used anymore (we use defaults from the zod parse). mostly left this in the repo for my own reference, but needs a cleanup

File: src/helpers/context/notebookContextRetrieval.ts [hunk 4]
```diff
@@ -11,13 +11,18 @@ export const STARTING_TEXTBOOK_CONTEXT: string =
...
+  jupyterbookURLs?: string[];
   attemptJupyterbookLinkExpansion?: boolean;
   debug?: boolean;
 }
 
+// Permit the plugin to start working before context is collected, if it takes longer than
+// this amount of time. It'll keep running context collection in the background, if it takes
+// longer.
+const SOFT_TIMEOUT = 5000;
+
```
Comment (timothyaveni) [inline src/helpers/context/notebookContextRetrieval.ts]: let the plugin load even before textbook context has loaded

File: src/helpers/markdown/extract-links-images.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
+import { Root } from 'remark-parse/lib';
+import { visit } from 'unist-util-visit';
+
+type Pos = {
+  start?: { line: number; column: number; offset?: number };
+  end?: { line: number; column: number; offset?: number };
+};
+
+type LinkLike = {
```
Comment (timothyaveni) [inline src/helpers/markdown/extract-links-images.ts]: mostly llm-generated. parse forward from the markdown source rather than 'backwards' through the rendered HTML

File: src/index.ts [hunk 6]
```diff
@@ -47,238 +44,165 @@ const plugin: JupyterFrontEndPlugin<void> = {
...
-        const [allCells, _, allowed_value] = parseNB(
-          notebook,
-          undefined,
-          finalConfig.activation_flag ?? '',
-          finalConfig.deactivation_flag ?? ''
-        );
-        allowed = allowed_value; // set this globally
+        await currentWidget.context.ready;
+        await currentWidget.revealed;
+
```
Comment (timothyaveni) [inline src/index.ts]: waiting for `context.ready` is a big part of resolving the race condition -- I don't think `.revealed` is really necessary now that we're not waiting on the render, but doesn't really hurt
